velox/expression/SignatureBinder.cpp

```diff
@@ -208,6 +208,11 @@ TypePtr SignatureBinder::tryResolveType(
     const std::unordered_map<std::string, std::string>& constraints) {
   const auto& params = typeSignature.parameters();
 
+  if (isDecimalName(typeSignature.baseType())) {
```

Comment:
What about combining this logic with `if (isCommonDecimalName(typeName)) { return inferDecimalType(...); }` below to return nullptr when precision and scale are not literal integers and are not bound by any constraint? (inferDecimalType() already has a check for this case, but throw an exception instead of returning nullptr.) The benefit would be that tryResolveType() could still resolve something like "short_decimal(5, 2)".

Reply:
`if (isCommonDecimalName(typeName)) `we enter this branch for type signature` DECIMAL(10, 5)` and not `SHORT_DECIMAL(10,5)`, hence the name `isCommonDecimalName`.

Like mentioned in the comment, today we don't support FunctionSignatures of the form:

`SHORT_DECIMAL(a, b)  fname(A, B)`

Also, I'm curious why this failure doesn't happen for Arithmetic functions, like you pointed out, DECIMA(a, b) signature would throw an exception.

Reply:
> if (isCommonDecimalName(typeName)) we enter this branch for type signature DECIMAL(10, 5) and not SHORT_DECIMAL(10,5), hence the name isCommonDecimalName.

I was thinking about making it `if (isCommonDecimalName(typeName) || isDecimalName(typeName)) { return inferDecimalType(...); }` and then make `inferDecimalType()` return nullptr instead of throwing an exception when precision and scale are unbound.

> Also, I'm curious why this failure doesn't happen for Arithmetic functions, like you pointed out, DECIMA(a, b) signature would throw an exception.

That's a good question. I think when we call a Decimal arithmetic function, we give SignatureBinder a list of actual argument types and call SignatureBinder::tryBind() to determine the variable values (e.g., a_precision). SignatureBinder::tryBind() doesn't call `inferDecimalType()` but handle the variables by itself, so it doesn't throw with DECIMAL(a, b).

The case we come across internally is different. The internal code calls the static SignatureBinder::tryResolveType(arg, {}) on each individual argument type in the function signature.

Reply:
I addressed this suggestion here https://github.com/facebookincubator/velox/pull/2544
The Signature changes in this PR have been moved as well.